crates/ty_python_semantic/src/types.rs

```diff
@@ -13907,6 +13904,54 @@ impl<'db> UnionType<'db> {
 
         ConstraintSet::from(sorted_self == other.normalized(db))
     }
+
+    /// Returns true if this union is equivalent to `int | float`, which is what `float` expands
+    /// into in type position.
+    pub(crate) fn is_int_float(self, db: &'db dyn Db) -> bool {
+        let elements = self.elements(db);
+        if elements.len() != 2 {
+            return false;
+        }
+        let mut has_int = false;
+        let mut has_float = false;
+        for element in elements {
+            if let Type::NominalInstance(nominal) = element
+                && let Some(known) = nominal.known_class(db)
+            {
+                match known {
+                    KnownClass::Int => has_int = true,
+                    KnownClass::Float => has_float = true,
+                    _ => {}
+                }
+            }
+        }
+        has_int && has_float
+    }
+
+    /// Returns true if this union is equivalent to `int | float | complex`, which is what
+    /// `complex` expands into in type position.
+    pub(crate) fn is_int_float_complex(self, db: &'db dyn Db) -> bool {
+        let elements = self.elements(db);
+        if elements.len() != 3 {
+            return false;
+        }
+        let mut has_int = false;
+        let mut has_float = false;
+        let mut has_complex = false;
+        for element in elements {
+            if let Type::NominalInstance(nominal) = element
+                && let Some(known) = nominal.known_class(db)
+            {
+                match known {
+                    KnownClass::Int => has_int = true,
+                    KnownClass::Float => has_float = true,
+                    KnownClass::Complex => has_complex = true,
+                    _ => {}
+                }
+            }
+        }
+        has_int && has_float && has_complex
+    }
```

Comment:
This could be a single method which returns an option of an enum, since these are mutually exclusive possibilities that require effectively the same work to check... that would be slightly less redundant and slightly more efficient? But I think it doesn't really matter, this way is totally fine.